src/cloud-search-index-storage.js

```diff
@@ -0,0 +1,210 @@
+  }
+
+  withLog(logger) {
+    this.log = logger;
+    return this;
+  }
+
+  /**
+   * Strong Existence Check
+   *
+   * Needs to have both the sourceId and the contentHash (source binary hash) matching
+   * @param {Object} query
+   * @returns Object of hits if there is a hit or undefined if it doesn't exist
+   */
+  async exists(query) {
+    const facetFilters = [];
+    // if no contentHash, we don't want to match just on the sourceId
+    // becuase that is not reliable so we will not return a match
+    if (!query.contentHash) {
+      console.log('Missing contentHash, no existence in the record storage.');
+      return undefined;
+    }
+
+    facetFilters.push(`contentHash:${query.contentHash}`);
+
+    if (query?.sourceAssetId) {
+      facetFilters.push(`sourceAssetId:${query.sourceAssetId}`);
+    }
+
+    const searchResult = await this.index.search(
+      '',
+      {
+        facetFilters,
+      },
+    );
+    if (searchResult?.nbHits > 0) {
+      return searchResult?.hits;
+    }
+    return false;
+  }
+
+  /**
+   * Retrieve a single index record via the record's identifier (objectID).
+   * @param {*} objectID - the unique identifier for the index record
+   * @returns the raw result of the query; will contain the record if it exists.
+   */
+  async get(objectID) {
+    return this.index.search(
+      '',
+      {
+        facetFilters: [
+          `objectID:${objectID}`,
+        ],
+      },
+    );
+  }
+
+  /**
+   * Save the information in the provided object to the search index as an index record.
+   *
+   * Not all fields provided in the object will be saved to the index. This method takes the values
+   * from the provided object that are specified to be used in the index and saves a record with
+   * those values; other values are ignored.
+   * @param {Object} doc the object to store in the index.
+   * @returns An object containing the saved record's ObjectID.
+   */
+  async save(doc) {
+    const indexRecord = Object.keys(doc).reduce((obj, key) => {
+      const result = clone(obj);
+      if (this.#FIELDS_TO_INDEX.has(key)) {
+        result[key] = doc[key];
+      }
+      return result;
+    }, {});
+    if (!indexRecord?.objectID) {
+      indexRecord.objectID = crypto.randomUUID();
+    }
+
+    this.log.info(`Saving doc to cloud record storage index ${this.indexName}`, indexRecord);
+    return this.index.saveObject(indexRecord);
+  }
+
+  /**
+   * Get a list of ObjectID that contain the contentHash
+   * @param {String} contentHash sha256 hash of the source binary
+   * @returns list of strings of ojectIDs
+   */
+  async getObjectIdsByContentHash(contentHash) {
+    const searchResult = await this.index.search(
+      '',
+      {
+        distinct: false,
+        facetFilters: [
+          `contentHash:${contentHash}`,
+        ],
+      },
+    );
+    if (searchResult?.nbHits > 0) {
+      return searchResult?.hits.map((hit) => hit.objectID);
+    }
+    return false;
```

Comment:
If this returns a list of objectIDs when matches are found, do you think it should return an empty list if there are no matches found?